server/src/query/mod.rs

```diff
@@ -0,0 +1,545 @@
+//! * [`SortedJoinExec`] which joins the sorted arrays produced by each [`GridExec`] and combines
+//! them with the time series tags retrieved from the
+//! [`MetadataManager`](crate::metadata::MetadataManager) to create the complete results containing
+//! a timestamp column, one or more field columns, and zero or more tag columns.
+
+// Public so the rules added to Apache Arrow DataFusion's physical optimizer can access GridExec.
+pub mod grid_exec;
+pub mod sorted_join_exec;
+
+use std::any::Any;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use datafusion::arrow::datatypes::{ArrowPrimitiveType, DataType, SchemaRef, TimeUnit};
+use datafusion::common::ToDFSchema;
+use datafusion::datasource::object_store::ObjectStoreUrl;
+use datafusion::datasource::{
+    datasource::TableProviderFilterPushDown, listing::PartitionedFile, TableProvider, TableType,
+};
+use datafusion::error::{DataFusionError, Result};
+use datafusion::execution::context::{ExecutionProps, SessionState};
+use datafusion::logical_expr::{self, BinaryExpr, Expr, Operator};
+use datafusion::optimizer::utils;
+use datafusion::physical_expr::planner;
+use datafusion::physical_plan::{
+    file_format::FileScanConfig, file_format::ParquetExec, ExecutionPlan, Statistics,
+};
+use object_store::ObjectStore;
+use tokio::sync::RwLockWriteGuard;
+
+use crate::metadata::model_table_metadata::ModelTableMetadata;
+use crate::query::grid_exec::GridExec;
+use crate::query::sorted_join_exec::{SortedJoinElement, SortedJoinExec};
+use crate::storage;
+use crate::storage::StorageEngine;
+use crate::types::ArrowValue;
+use crate::Context;
+
+/// A queryable representation of a model table which stores multivariate time
+/// series as segments containing metadata and models. [`ModelTable`] implements
+/// [`TableProvider`] so it can be registered with Apache Arrow DataFusion and
+/// the multivariate time series queried as multiple univariate time series.
+pub struct ModelTable {
+    /// Access to the system's configuration and components.
+    context: Arc<Context>,
+    /// Location of the object store used by the storage engine.
+    object_store_url: ObjectStoreUrl,
+    /// Metadata required to read from and write to the model table.
+    model_table_metadata: Arc<ModelTableMetadata>,
+    /// Field column to use for queries that do not include fields.
+    fallback_field_column: u16,
+}
+
+impl ModelTable {
+    pub fn new(context: Arc<Context>, model_table_metadata: Arc<ModelTableMetadata>) -> Arc<Self> {
+        // Compute the index of the first field column in the model table's
+        // schema. This is used for queries that does not contain any fields.
+        let fallback_field_column = {
+            model_table_metadata
+                .schema
+                .fields()
+                .iter()
+                .position(|field| field.data_type() == &ArrowValue::DATA_TYPE)
+                .unwrap() as u16 // unwrap() is safe as model tables contains fields.
```

Comment:
Maybe something like `unwrap() is safe as all model tables contain at least one field.` instead. There might also be some confusion between schema fields and model table fields here.

Reply:
Fixed by using your suggested comment.